central/externalbackups/plugins/s3compatible/s3compatible.go

```diff
@@ -0,0 +1,231 @@
+)
+
+var (
+	log = logging.LoggerForModule()
+)
+
+func init() {
+	plugins.Add(types.S3CompatibleType, func(backup *storage.ExternalBackup) (types.ExternalBackup, error) {
+		return newS3Compatible(backup)
+	})
+}
+
+type s3Compatible struct {
+	integration *storage.ExternalBackup
+
+	client   *s3.Client
+	uploader *manager.Uploader
+}
+
+func validate(cfg *storage.S3Compatible) error {
+	errorList := errorhelpers.NewErrorList("S3 Compatible Validation")
+	if cfg.GetAccessKeyId() == "" {
+		errorList.AddString("Access Key ID must be specified")
+	}
+	if cfg.GetSecretAccessKey() == "" {
+		errorList.AddString("Secret Access Key must be specified")
+	}
+	if cfg.GetRegion() == "" {
+		errorList.AddString("Region must be specified")
+	}
+
+	return errorList.ToError()
+}
+
+func validateEndpoint(endpoint string) error {
+	if _, err := url.Parse(fmt.Sprintf("https://%s", urlfmt.TrimHTTPPrefixes(endpoint))); err != nil {
+		return errors.Wrapf(err, "invalid URL %s", endpoint)
+	}
+	return nil
+}
+
+func newS3Compatible(integration *storage.ExternalBackup) (*s3Compatible, error) {
+	cfg, ok := integration.GetConfig().(*storage.ExternalBackup_S3Compatible)
+	if !ok {
+		return nil, errors.New("S3 Compatible configuration required")
+	}
+	if err := validate(cfg.S3Compatible); err != nil {
+		return nil, err
+	}
+
+	opts := []func(*config.LoadOptions) error{
+		config.WithRegion(cfg.S3Compatible.GetRegion()),
+		config.WithHTTPClient(&http.Client{Transport: proxy.RoundTripper()}),
+		config.WithCredentialsProvider(
+			credentials.NewStaticCredentialsProvider(
+				cfg.S3Compatible.GetAccessKeyId(),
+				cfg.S3Compatible.GetSecretAccessKey(), "",
+			),
+		),
+	}
+
+	ctx, cancelFn := context.WithTimeout(context.Background(), initialConfigurationMaxTimeout)
+	defer cancelFn()
+	awsConfig, err := config.LoadDefaultConfig(ctx, opts...)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to load the aws config")
+	}
+
+	var clientOpts []func(*s3.Options)
+	if cfg.S3Compatible.GetUrlStyle() == storage.S3URLStyle_S3_URL_STYLE_PATH {
+		clientOpts = append(clientOpts, func(o *s3.Options) {
+			o.UsePathStyle = true
+		})
+	}
+
+	if endpoint := cfg.S3Compatible.GetEndpoint(); endpoint != "" {
+		if err := validateEndpoint(endpoint); err != nil {
+			return nil, err
+		}
+		clientOpts = append(clientOpts, func(o *s3.Options) {
+			o.BaseEndpoint = aws.String(endpoint)
+		})
+	}
+
+	client := s3.NewFromConfig(awsConfig, clientOpts...)
+	return &s3Compatible{
+		integration: integration,
+		client:      client,
+		uploader:    manager.NewUploader(client),
+	}, nil
+}
+
+func (s *s3Compatible) Backup(reader io.ReadCloser) error {
+	defer func() {
+		if err := reader.Close(); err != nil {
+			log.Errorf("closing reader: %+v", err)
+		}
+	}()
+
+	log.Info("Starting S3 Compatible Backup")
+	ctx, cancelFn := context.WithTimeout(context.Background(), backupMaxTimeout)
+	defer cancelFn()
+	formattedTime := time.Now().Format("2006-01-02T15:04:05")
```

Comment:
nit: could pull the format string into `const`.

Also, you can be really fancy and do `backup_2006_01_02_15_04_05.zip` as the format string, then you don't need to manually do the `fmt.Sprintf` call below and have this be done by the `Format()`.

Reply:
> you can be really fancy and do backup_2006_01_02_15_04_05.zip as the format string, then you don't need to manually do the fmt.Sprintf call below and have this be done by the Format()

TIL ☝️  and done 🙂